src/Fields/mapreduce_cuda.jl

```diff
@@ -0,0 +1,236 @@
+local_sum_cuda(
+    field::Field{V},
+) where {
+    Nij,
+    A <: AbstractArray,
+    V <:
+    Union{DataLayouts.IJFH{<:Any, Nij, A}, DataLayouts.VIJFH{<:Any, Nij, A}},
+} = mapreduce_cuda(identity, +, field, weighting = true)
+
+local_maximum_cuda(
+    fn,
+    field::Field{V},
+) where {
+    Nij,
+    A <: AbstractArray,
+    V <:
+    Union{DataLayouts.IJFH{<:Any, Nij, A}, DataLayouts.VIJFH{<:Any, Nij, A}},
+} = mapreduce_cuda(fn, max, field)
+
+local_minimum_cuda(
+    fn,
+    field::Field{V},
+) where {
+    Nij,
+    A <: AbstractArray,
+    V <:
+    Union{DataLayouts.IJFH{<:Any, Nij, A}, DataLayouts.VIJFH{<:Any, Nij, A}},
+} = mapreduce_cuda(fn, min, field)
+
+function mapreduce_cuda(
+    f,
+    op,
+    field::Field{V};
+    weighting = false,
+    opargs...,
+) where {
+    Nij,
+    A <: AbstractArray,
+    V <:
+    Union{DataLayouts.IJFH{<:Any, Nij, A}, DataLayouts.VIJFH{<:Any, Nij, A}},
+}
+    data = Fields.field_values(field)
+    pdata = parent(data)
+    T = eltype(pdata)
+    (_, _, _, Nv, Nh) = size(data)
+    Nf = div(length(pdata), prod(size(data))) # length of field dimension
+    wt = Spaces.weighted_jacobian(axes(field)) # wt always IJFH layout
+    pwt = parent(wt)
+
+    nitems = Nv * Nij * Nij * Nh
+    max_threads = 256# 512 1024
+    nthreads = min(max_threads, nitems)
+    # perform n ops during loading to shmem
+    n_ops_on_load = cld(nitems, nthreads) == 1 ? 0 : 7
+    effective_blksize = nthreads * (n_ops_on_load + 1)
+    nblocks = cld(nitems, effective_blksize)
+
+    reduce_cuda = CuArray{T}(undef, nblocks, Nf)
+    reduce_cuda_fld = CuArray{T}(undef, Nf)
+    shmemsize = nthreads
+    # place each field on a different block
+    @cuda threads = (nthreads) blocks = (nblocks, Nf) shmem =
+        shmemsize * sizeof(T) mapreduce_cuda_kernel!(
```

Comment:
I think `shmem` is only for dynamic shared memory: we should be able to use static here.
```suggestion
    @cuda threads = (nthreads) blocks = (nblocks, Nf) mapreduce_cuda_kernel!(
```

Reply:
Or we use dynamic shared memory and don't `Val`-specialize it?